src/chain.jl

```diff
@@ -0,0 +1,37 @@
+
+
+# Some experiments with chain to start removing the need for recur to be mutable.
+# As per the conversation in the recurrent network rework issue.
+
+# Main difference between this and the _applychain function is we return a new chain
+# with the internal state modified as well as the output of applying x to the chain.
+apply(chain::Flux.Chain, x) = begin
+  layers, out = apply(chain.layers, x)
+  Flux.Chain(layers), out
+end
+
+apply(layers::NamedTuple{NMS, TPS}, x) where {NMS, TPS} = begin
+  layers, out = apply(Tuple(layers), x)
+  NamedTuple{NMS}(layers), out
+end
+
+function apply(layers::AbstractVector, x)  # type-unstable path, helps compile times
+  new_layers = typeof(layers)(undef, length(layers))
+  for (idx, f) in enumerate(layers)
+    new_layers[idx], x = apply(f, x)
+  end
```

Comment:
This won't be differentiable, unfortunately. I don't know if there are any non-hacky tricks to make it work. We might have to define our own `scan` function or something like it.

Reply:
Ok. Yeah, I tried a few other options, but I think a `_scan` function with a hand written rrule is likely what we need. I'm not too adept at writing rrules yet, but will work on it. In the mean time I've added some tests for checking gradient consistency and the scan function with a placeholder for the rrule and skip the gradient test for the vector version.